Everywhere: Implement synchronization primitives for win32

R-Goc opened the pull request:
The goal of this PR is to remove the dependency on pthread outside of LibThreading. Pthreads are an implementation detail and we shouldn't be relying on them everywhere. This PR is the first part towards this. Synchronization APIs are abstracted away and added where needed. This will solve the biggest issue with pthreads-win32 which is lack of interprocess synchronization primitives. Future PRs will move further with this, but as this is very conflict prone I decided to split it up.

R-Goc committed:
LibThreading: Add Once.h

Once.h provides a way to run a function once without relying on system
libraries at the call site. It also allows use of any function as long
as it satisfies the VoidFunction concept. For similarity with the STL I
would have preferred it to be in AK, but it needs a mutex to work so
that would cause a circular dependency.
With some templates and an optional it could be modified to return non
void types, but since that isn't an existing pattern in the codebase
that could be implemented in a follow-up.
Performance wise compared to pthread_once it should be very similar
in the already initialized case as both use an atomic. On the first call
this might be slightly slower since we use a mutex while pthread may be
doing something fancier with a futex or similar, but it shouldn't
matter.

R-Goc committed:
LibCore: Use call_once from LibThreading

This commit makes ThreadEventQueue use call_once instead of
pthread_once_t. The aim is to limit the pthread dependency to
LibThreading.

R-Goc commented on Libraries/LibThreading/Once.h:
```diff
@@ -0,0 +1,36 @@
+            return;
+
+        callable();
+        flag.has_been_called.store(true, MemoryOrder::memory_order_relaxed);
```
I was thinking about this. The second load can be relaxed for sure. I think I'll make the store release ordering, even if it is just for the compiler fence. Not entirely sure if it necessary with the mutex in place.

gmta replied:
It's needed for threads that don't take the mutex at all I believe

R-Goc committed:
LibThreading/LibSync: Split out sync primitives

This commit splits out synchronization primitives from LibThreading into
LibSync. This is because LibThreading depends on LibCore, while LibCore
needs the synchronization primitives from LibThreading. This worked
while they were header only, but when I tried to add an implementation
file it ran into the circular dependency. To abstract away the pthread
implementation using cpp files is necessary so the synchronization
primitives were moved to a separate library.

R-Goc committed:
LibSync: Abstract away Mutex implementation

This commit adds in-place pimpl to abstract away the implementation of
Mutex. It also adds policies to configure the type of mutex desired.

Because of the tight integration between mutex and condition variables
they also needed to be reworked and the changes have to be in one commit
to retain atomicity.

A win32 and pthread implemenation is provided to make sure the api
works with both.

R-Goc commented on Libraries/LibThreading/Once.h:
Actually no, this is wrong. As set_up() can take time. So the cas will make is_set_up true before set_up() finished running which could be very bad.

R-Goc commented:
Unstale. I'll get to this soon-ish.

ADKaster commented on Libraries/LibThreading/Thread.cpp:
```diff
@@ -5,9 +5,15 @@
 
 namespace Threading {
 
+bool ThreadID::operator==(ThreadID const& other)
```
this snippet feels like it should be moved to this commit in some way: UI/Qt/EventLoopImplementationQt.cpp:

```c++
#if defined(AK_OS_WINDOWS)
namespace AK {

template<>
struct Traits<pthread_t> : public DefaultTraits<pthread_t> {
    static unsigned hash(pthread_t thread_id)
    {
        return Traits<intptr_t>::hash(reinterpret_cast<intptr_t>(thread_id.p));
    }

    static constexpr bool equals(pthread_t a, pthread_t b)
    {
        return Traits<intptr_t>::equals(reinterpret_cast<intptr_t>(a.p), reinterpret_cast<intptr_t>(b.p));
    }
};

}

#endif
```

ADKaster commented on Libraries/LibSync/ConditionVariable.h:
```diff
@@ -1,63 +1,73 @@
-private:
-    pthread_cond_t m_condition;
-    Mutex& m_to_wait_on;
+    alignas(void*) unsigned char m_storage[get_storage_size()];
```
```suggestion
#ifdef AK_OS_WINDOWS
    using StorageType = void*;
#else
    using StorageType = pthread_cond_t;
    
    alignas(StorageType) unsigned char m_storage[sizeof(StorageType)];
    MutexType& m_to_wait_on;
```

R-Goc commented on Libraries/LibSync/MutexPosix.cpp:
```diff
@@ -6,25 +6,86 @@
+template<typename R, typename I>
+MutexBase<R, I>::~MutexBase()
 {
+    int result = pthread_mutex_destroy(reinterpret_cast<pthread_mutex_t*>(m_storage));
```
Changed these to an internal function where it made sense. In the mutex implementation on windows we have 3 different types so I didn't do it there.

R-Goc commented on Libraries/LibCore/ThreadEventQueue.cpp:
```diff
@@ -46,11 +47,11 @@ struct ThreadEventQueue::Private {
 ThreadEventQueue& ThreadEventQueue::current()
 {
-    pthread_once(&s_current_thread_event_queue_key_once, [] {
+    call_once(s_current_thread_event_queue_key_once, [] {
```
After the changes made here it becomes more complicated to do this and keep the current_or_null. I'll leave this for a future PR.

R-Goc commented on Libraries/LibIPC/CMakeLists.txt:
```diff
@@ -18,4 +18,4 @@ else()
 
 ladybird_lib(LibIPC ipc)
-target_link_libraries(LibIPC PRIVATE LibCore LibURL LibThreading)
+target_link_libraries(LibIPC PRIVATE LibCore LibURL LibThreading LibSync)
```
Looks like it, I will move it.

R-Goc committed:
LibSync: Remove pthread from the RWLock interface

This commit removes any use of pthread.h from the RWLock header. It also
adds a win32 implementation of RWLock.

Support for interprocess RWLocks is left to a later PR.

Due to the API design of SRWLock the unlock methods must match the lock
method.
